### PR TITLE
ggml : move FP16 <-> FP32 stuff to ggml-impl.h

### LINKED ISSUE #3858: Speed drops since b1442
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x ] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x ] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering- and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x ] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

Speed was typical 6.3 tokens/sec for Win10, i5850, 32 GB both with supplied main.exe AVX2 up to build 1440
Test confirmed by GCC compiled under CodeBlocks.

# Current Behavior

Beginning with Build b1442 (confirmed with b1443, b1444, b1445) speed drops to 2.84 tokens / second under same
test conditions.

# Environment and Context

Test done at equal conditions with the following start:

main -t 6 -m models/META-2/7B/META_2_7B_f16_V4_q4_0.gguf --log-disable --color -c 4096 --temp 0.7 --repeat_penalty 1.1 -n -1 --in-prefix-bos --in-prefix [INST] --in-suffix [/INST] -i -p "[INST] <<SYS>> You are a helpful, respectful and honest assistant. <</SYS>> Write Good Morning in German. [/INST]"

After the complete answer manually stopped by Ctrl C and speed token/sec taken.

* Physical (or virtual) hardware you are using, e.g. for Linux:

` i5 8350, 32 GB RAM, 1TB M.2 SSD

* Operating System, e.g. for Linux:

`WIN 10, up to date

* SDK version, e.g. for Linux:
*Checked CPU only version AVX2 packages: e.g. llama-b1440-bin-win-avx2-x64.zip

Checked with provided b1440, b1442 ... b1445 AVX2 packages (main.exe and  dll's from package)
Counter checked with GCC, IDE is CodeBlocks build1440 performs as expected, build 1445 same speed drop as referenced versions

# Steps to Reproduce

Please provide detailed steps for reproducing the issue. We are not sitting in front of your screen, so the more detail the better.
1. step 1 Unpack build AVX2  dll's and main  into clean test environment (root directory drive C:\LLaMa2\0_KI_test)
2. step 2 start main by batch file
3. step 3 Wait till system completes response.
4. step 4 Hit Ctrl C, confirm Yes and take note of tokens/sec 

Notice speed drop after build 1440 (starting b1442)  b1440:
llama-b1440-bin-win-avx2-x64.zip  6.3 tokens/sec,
llama-b1442-bin-win-avx2-x64.zip  2.84 tokens/sec
llama-b1443-bin-win-avx2-x64.zip  2.84 tokens/sec
llama-b1444-bin-win-avx2-x64.zip  2.84 tokens/sec
llama-b1445-bin-win-avx2-x64.zip  2.84 tokens/sec

### PR DESCRIPTION
**Pull request by ggerganov**
close #3858 

Alternative approach to fix FP16 <-> FP32 performance degradation due to #3833 

Introduce `ggml-impl.h` and move common stuff (mostly macros) from `ggml.c` into it for reuse in `ggml-quants.c`

### CODE REVIEW

**Comment by ggerganov** [comment]
Any ideas how to fix the Windows CI? I guess something fails with the initialization of the `table_f32_f16` symbol but don't have Windows to test

**Comment by slaren** [comment]
This fixes the issue under Windows with MSVC:

| model                          |       size |     params | backend    |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | CPU        |          8 | tg 32      |     13.65 ± 0.16 |

build: 223696c (1449)


| model                          |       size |     params | backend    |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | CPU        |          8 | tg 32      |     10.71 ± 0.39 |

build: 6e08281 (1445)

| model                          |       size |     params | backend    |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | CPU        |          8 | tg 32      |     13.04 ± 0.25 |

build: ff3bad8 (1441)

**Comment by slaren** [comment]
On a side note, I wasted a lot of time realizing that the default cmake build options under Windows no longer include AVX. This seems to be because `LLAMA_NATIVE` is enabled by default, which disables AVX, but doesn't work with MSVC.

**Comment by cebtenzzre** [comment]
> On a side note, I wasted a lot of time realizing that the default cmake build options under Windows no longer include AVX. This seems to be because `LLAMA_NATIVE` is enabled by default, which disables AVX, but doesn't work with MSVC.

#809 would fix this.

File: ggml-impl.h [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+#define GGML_FP16_TO_FP32(x) ggml_lookup_fp16_to_fp32(x)
+#define GGML_FP32_TO_FP16(x) GGML_COMPUTE_FP32_TO_FP16(x)
+
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Comment (slaren) [inline ggml-impl.h:231]: This header may be included in C++ code such as `ggml-cuda.cu` or `ggml-opencl.cpp`, and it contains `extern` symbols, so this should probably be moved to the start of the file already.

**Review by slaren — APPROVED** [review]

**Pull request merged**